### PR TITLE
Reduce flakiness in the render tests

### PR DESCRIPTION
**Pull request by stepankuzmin**
Most of the flakiness happens because fetch throws a network error during assets request (e.g., raster/vector tiles or geojson data).

https://github.com/mapbox/mapbox-gl-js/blob/a65da974c5578f4b79f9a64c73a2973980a99275/src/util/ajax.js#L155-L161

This PR switches to the `testem` built-in `express` server to serve assets in the render tests and enhance asset caching.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -159,6 +159,16 @@ defaults: &defaults
     - image: cimg/node:14.16-browsers
   working_directory: ~/mapbox-gl-js
 
+use_ramdisk: &use_ramdisk
+  working_directory: /mnt/ramdisk/mapbox-gl-js
```
Comment (SnailBones) [inline .circleci/config.yml:163]: Did you notice more consistent test running times with the ramdisk? Is it also possible to implement this on Mac and Windows?

Comment (mourner) [reply .circleci/config.yml:163]: Here's an article about using a RAM Disk in Windows, a little more involved but worth a try https://support.circleci.com/hc/en-us/articles/4411520952091-Create-a-windows-RAM-disk

Comment (stepankuzmin) [reply .circleci/config.yml:163]: It used to be a bit faster when I tried it before adding Windows and Mac, but now it's not so obvious, so we can also consider dropping this and not adding extra complexity.

Comment (SnailBones) [reply .circleci/config.yml:163]: Thoughts on isolating it in a separate PR to better compare changes and not block the other improvements here?

Comment (stepankuzmin) [reply .circleci/config.yml:163]: Sure, let's first land a flakiness fix and experiment with the RAM disk after.

File: patches/testem+3.10.0.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/node_modules/testem/lib/server/index.js b/node_modules/testem/lib/server/index.js
+index a1baf70..e3545f7 100644
+--- a/node_modules/testem/lib/server/index.js
++++ b/node_modules/testem/lib/server/index.js
+@@ -177,7 +177,7 @@ class Server extends EventEmitter {
+       }
+       next();
+     });
+-    app.use(express.static(`${__dirname}/../../public`));
++    app.use('/testem', express.static(`${__dirname}/../../public/testem`, {fallthrough: false}));
```
Comment (stepankuzmin) [inline patches/testem+3.10.0.patch:10]: Testem uses an express [`serve-static`](https://github.com/expressjs/serve-static) middleware to serve the assets. It's mounted on the app root, meaning every request is first processed with the `express.static` middleware and only after that with the middlewares defined in the config, e.g., proxy middleware.

This ends up in a bunch of extra `fs.stat` calls which bloat the server response time. You can check that by enabling the `DEBUG=send` environment variable and making a request to the server.

I made a PR that fixes this behavior in the `testem` https://github.com/testem/testem/pull/1615

File: test/integration/lib/render.js [hunk 3]
```diff
@@ -147,16 +148,17 @@ async function setupLayout(options) {
...
     for (const prop in currentFixture) {
         if (prop.indexOf('expected') > -1) {
-            let path = `${currentTestName}/${prop}.png`;
-            // regression tests with # in the name need to be sanitized
-            path = encodeURIComponent(path);
+            const path = `/${currentTestName}/${prop}.png`
+                .split('/')
+                // regression tests with # in the name need to be sanitized
+                .map(p => encodeURIComponent(p))
+                .join('/');
```
Comment (SnailBones) [inline test/integration/lib/render.js:160]: @stepankuzmin Could you share context for this change? Was there a performance issue caused by encoding full paths?

File: package.json [hunk 4]
```diff
@@ -88,9 +90,9 @@
     "rollup-plugin-terser": "^7.0.2",
     "rollup-plugin-unassert": "^0.5.0",
     "selenium-webdriver": "^4.4.0",
+    "serve-static": "^1.15.0",
     "shuffle-seed": "^1.1.6",
     "sinon": "^15.0.0",
-    "st": "^3.0.0",
```
Comment (mourner) [inline package.json:93]: `st` is still used in `start-server` script and consequently other scripts like `start` — let's either keep it, or replace the static server in the script with `serve-static` (CLI equivalent with same settings).

File: package.json [hunk 5]
```diff
@@ -144,7 +146,8 @@
     "test-style-spec": "cd src/style-spec && npm test",
     "prepublishOnly": "run-s build-flow-types build-dev build-prod-min build-prod build-csp build-css build-style-spec",
     "print-release-url": "node build/print-release-url.js",
-    "codegen": "build/run-node build/generate-style-code.js && build/run-node build/generate-struct-arrays.js"
+    "codegen": "build/run-node build/generate-style-code.js && build/run-node build/generate-struct-arrays.js",
+    "postinstall": "patch-package"
```
Comment (mourner) [inline package.json:150]: As discussed in chat, this may affect downstream users of `mapbox-gl` package, so let's do the patch as a part of CI rather than `install` in general.

File: src/util/ajax.js [hunk 6]
```diff
@@ -148,16 +148,15 @@ function makeFetchRequest(requestParameters: RequestParameters, callback: Respon
             if (response.ok) {
                 const cacheableResponse = cacheIgnoringSearch ? response.clone() : null;
                 return finishRequest(response, cacheableResponse, requestTime);
-
             } else {
                 return callback(new AJAXError(response.statusText, response.status, requestParameters.url));
             }
         }).catch(error => {
-            if (error.code === 20) {
+            if (error.name === 'AbortError') {
```
Comment (mourner) [inline src/util/ajax.js:155]: Nit: any reason for this change? Can an abort error be other than code 20, or is this just for clarity?

**Review by mourner — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (stepankuzmin) [reply package.json:93]: There is no CLI for `serve-static,` so I'll keep `st`

File: src/util/ajax.js (hunk 6 above)
Comment (stepankuzmin) [reply src/util/ajax.js:155]: Yep, it's just for the clarity

File: test/integration/lib/render.js (hunk 3 above)
Comment (stepankuzmin) [reply test/integration/lib/render.js:160]: When we encode the whole URL path with `encodeURIComponent`, it's being served with the `testem`'s built-in server and not our `serve-static` middleware. It might be okay since it's unnecessary to cache the `expectend.png` in the browser, but I think it's better to serve it with our server to be consistent with the other assets.

https://github.com/mapbox/mapbox-gl-js/blob/c861dec0d4b2b78734881a4d6e64baf5d9b54892/test/integration/lib/middlewares.cjs#L22-L24

File: package.json (hunk 5 above)
Comment (stepankuzmin) [reply package.json]: I've dropped `postinstall` after the `testem` update

File: patches/testem+3.10.0.patch (hunk 2 above)
Comment (stepankuzmin) [reply patches/testem+3.10.0.patch]: I've removed the patch and updated `testem` to v3.10.1 which includes the fix.
